### ISSUE TITLE
Can't check for Android permission rationale required

### ISSUE THREAD

**Issue by easazade** [Baseflow/flutter-geolocator#994]
## 🐛 Bug Report
**TLDR**
there is no way provided in `geolocator` plugin to check if android permission rationale is required. since when permission rationale is required android OS does not allow request permission dialog to be shown by the app when we call method `Geolocator.requestPermission()`

**PROBLEM**
on android by default location permission status is `LocationPermission.denied` which is fine. but if we try to ask user to grant location permission using method `Geolocator.requestPermission()` and user chooses the options `deny` more than once or twice or choose option don't ask again then `Geolocator.requestPermission()` method won't work and won't show the request permission dialog which is still fine and how it should be. but the status of the LocationPermission on android OS will change to `permission rationale required`, which is when Android OS demands the app to provide user with some reason as why we are keep asking for the permission. at this point we should show user some dialog about why we need these permission then navigate user to app settings or location settings in our case so user grants the location permission since `Geolocator.requestPermission()` won't show the request permission dialog anymore. **THE PROBLEM IS** there is no way provided in `geolocator` plugin to check if android permission rationale is required. since when permission rationale is required android OS does not allow request permission dialog to be shown by the app when we call method `Geolocator.requestPermission()`

### Expected behavior
please provide a method to check if permission rationale is required on Android
### Reproduction steps
call `Geolocator.requestPermission()` a few times
### Configuration

**Version:** ^8.2.0

**Platform:** 
- [ x] :robot: Android

**Comment by easazade** [comment]
I checked again with [permission_handler](https://pub.dev/packages/permission_handler) plugin and the problem is not permission rationale. but there is something odd. when we use `Geolocator.requestPermission()` method and result `LocationPermission.deniedForever` the `Geolocator.checkPermission()` method will not return the same result and returns `LocationPermission.denied` instead. I'll open another issue

**Comment by mvanbeusekom** [comment]
Hi @easazade,

Once the user denied the permissions multiple times (Android API 30 and higher) or selected "Don't ask again" (Android API 29 and below) the `Geolocator.requestPermission()` will return `LocationPermission.deniedForever ` instead of `LocationPermission.denied`. 

According to the [Android documentation](https://developer.android.com/training/permissions/requesting#explain), the first time (and on Android API 29 and below also subsequent times) the requesting app receives a `denied` permission you should show the end user the permission rationale dialog. This is also the only time the `ActivityCompat.shouldShowRequestPermissionRationale()` ([API reference documentation](https://developer.android.com/reference/androidx/core/app/ActivityCompat#shouldShowRequestPermissionRationale(android.app.Activity,%20java.lang.String))) returns `true`. 

Meaning once the `Geolocator.requestPermission()` method returns `LocationPermission.permanentlyDenied` the App should redirect the user to the general location settings and the user should manually enable permissions for the App. The App can safely call the `Geolocator.requestPermission()` multiple times, it will always return `LocationPermission.deniedForever` and a permission dialog will not be shown. In this case there is no demand from Google to show a permission rationale to the user (although it doesn't hurt to explain the user in detail why the App needs permission and overing the user an option to navigate to the location settings via the `Geolocator.openLocationSettings()` method.

To summarize, the `Geolocator.requestPermission()` method can return two results when a user denies permissions:

1. `LocationPermission.denied`: permissions to access the device's location have been denied by the user. However the App is allowed to request permission again. The Android guidelines state that before requesting permissions again the App should first display a permission rationale to the user explaining in detail why the App requires location permission. Displaying the permission rationale is in no way monitored by the Android system and it can be skipped, this however might result in the App being denied when uploading to the Google Play Store;
2. `LocationPermission.deniedForever`: permissions to access the device's location have been permanently denied by the user. On Android API 29 and below this happens when the user selected the "Don't ask again" option, from Android API 30 and higher this happens when the user denied permissions two times in a row. After the App receives the `LocationPermission.permanentlyDenied` the only way to allow the App to access the device's location would be if the user grants permission through the location settings.

**Comment by mvanbeusekom** [comment]
@easazade no need to open another issue. The `Geolocator.checkPermission()` method cannot distinguish between `LocationPermission.denied` or `LocationPermission.deniedForever` as the Android SDK only supports returning a `GRANTED` or `DENIED` state when checking permissions. Android itself doesn't distinguish between one time denied and denied forever.

This topic is explained in more detail on our [Wiki page](https://github.com/Baseflow/flutter-geolocator/wiki/Breaking-changes-in-7.0.0).

**Comment by easazade** [comment]
thank you for the fast response

**Comment by darkstarx** [comment]
That was a mistake. You just had to expose the method `shouldShowRequestPermissionRationale` for every platform (on non-Android platforms just return true if denied) and allow users of the plugin to decide themselves how to work with this functionality. That's because only Android OS knows when `shouldShowRequestPermissionRationale` returns true and false, including the case 'Ask every time'.

**Comment by mvanbeusekom** [comment]
Hi @darkstarx thank you for your feedback. However I fail to see how `shouldShowPermissionRationale` will help out in this situation.

In cases where the user allowed permission once, each time the application starts the `shouldShowPermissionRationale` will return `false` so using this method will not give you any additional trigger to request permissions.

So when a user initially permanently denies permissions, closes the app and changes permissions in settings to "Ask every time ". There is no indication you are allowed to request permissions again (`shouldShowPermissionRationale` will still return `false`and only returns `true` after you requested permission and the user denied.

I would be happy to hear constructive alternatives or feedback.

**Comment by darkstarx** [comment]
The reason is quite simple. I want to call this method before requesting permissions. In other words, I'd like to know the status and not to show the system request dialog. In current implementation if the requesting is possible the system dialog will appear on requesting permissions. But I need to know the status (`denied` or `deniedForever`) silently for the user. And if it was `denied`, I'd show the dialog with rationale and button Request, but if it was `deniedForever`, I wouldn't show this dialog because the button Request won't help to get permission. Can you see the reason?
This means that I need to check permissions (this is currently possible with `Geolocator.checkPermission()`) and then to check the result of `shouldShowRequestPermissionRationale` (this is't possible in the current version of the plugin). Thus, you use the `shouldShowRequestPermissionRationale` in the `Geolocator.requestPermission()` but don't use it in the `Geolocator.checkPermission()`. And that was your choice, but not the choice of the plugin user. That's why I wrote that it was a mistake to hide this method. You just don't allow me to use full functionality of the platform.

Generally, it's always a mistake to hide functions which could actually be available in any module or plugin, just because you can't know which part of the module would be used by the module client. If you hide something, the module loses flexibility of usage.

**Comment by mvanbeusekom** [comment]
The reason we left this method out was to have the biggest common cross-platform abstraction. I am afraid that adding the method (and have it return `true` on all other platforms) causes more confusion then it would solve having it available and only working (in a already confusing way) on Android.

The use case you suggest would also not work. The outcome of the `checkPermission` doesn't affect the result of the `shouldShowRequestPermissionRationale` unfortunately. Meaning it is still not a reliable way to determine if permission was originally "denied" or "denied for ever". This is why we decided the usage value is minimal.

With the current implementation it is relative easy to accomplish the same behaviour by storing a boolean in shared storage after requesting permissions. If after you request permission the result is `denied` you set the boolean to `true` in all other cases you set it to `false`.

In a future version of the geolocator, I am planning to offer a one-to-one copy of the platform APIs next to a cross-platform API allowing developers to pick which suits them best. However this is still in development and will take some time.

**Comment by darkstarx** [comment]
> The outcome of the `checkPermission` doesn't affect the result of the `shouldShowRequestPermissionRationale` unfortunately.

Of course it shouldn't affect. This methods should be separate. The `checkPermission` gives me the answer if I should ask the user to allow the access, and the `shouldShowRequestPermissionRationale` gives the answer if the platform will actually show the system request dialog if I ask the user to allow the access. In other words, the second method might have name `shouldRequestPermission`.

> With the current implementation it is relative easy to accomplish the same behaviour by storing a boolean in shared storage after requesting permissions. If after you request permission the result is denied you set the boolean to true in all other cases you set it to false.

That would mimic the platform behavior which might be changed in some next version of the OS and could lead to some troubles, so it's a bad way. The only right way is to use platform functions when they exist.

> In a future version of the geolocator, I am planning to offer a one-to-one copy of the platform APIs next to a cross-platform API allowing developers to pick which suits them best. However this is still in development and will take some time.

It would be great!

Thanks anyway.
